File: sdk/java/src/main/java/com/vireo/WireFormat.java
package com.vireo;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.*;

/**
 * Vireo Wire Format Serialization
 * 
 * Binary canonical format for Vireo messages
 */
public class WireFormat {

    private static final int WIRE_MAGIC = 0x56495245; // 'VIRE'
    private static final int WIRE_VERSION = 0x03000000; // v3.0.0
    private static final int WIRE_HEADER_SIZE = 16;

    private static final ObjectMapper objectMapper = new ObjectMapper()
            .registerModule(new JavaTimeModule())
            .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

    private static final Map<Types.MessageType, Integer> TYPE_MAP = new HashMap<>();
    private static final Map<Integer, Types.MessageType> TYPE_REVERSE_MAP = new HashMap<>();

    static {
        TYPE_MAP.put(Types.MessageType.PROPOSE, 1);
        TYPE_MAP.put(Types.MessageType.COMMIT, 2);
        TYPE_MAP.put(Types.MessageType.EXECUTE, 3);
        TYPE_MAP.put(Types.MessageType.VERIFY, 4);
        TYPE_MAP.put(Types.MessageType.ESCALATE, 5);
        TYPE_MAP.put(Types.MessageType.DONE, 6);
        TYPE_MAP.put(Types.MessageType.FAILED, 7);
        TYPE_MAP.put(Types.MessageType.TIMEOUT, 8);
        TYPE_MAP.put(Types.MessageType.ACK, 9);
        TYPE_MAP.put(Types.MessageType.NACK, 10);
        TYPE_MAP.put(Types.MessageType.QUERY, 11);
        TYPE_MAP.put(Types.MessageType.RESPONSE, 12);
        TYPE_MAP.put(Types.MessageType.ERROR, 13);

        for (Map.Entry<Types.MessageType, Integer> entry : TYPE_MAP.entrySet()) {
            TYPE_REVERSE_MAP.put(entry.getValue(), entry.getKey());
        }
    }

    /**
     * Serialize a message envelope to binary format
     */
    public static byte[] serialize(Types.MessageEnvelope envelope) throws Errors.SerializationException {
        try {
            byte[] headerBytes = serializeHeader(envelope.getHeader());
            byte[] bodyBytes = envelope.getBody() != null ? serializeBody(envelope.getBody()) : new byte[0];

            byte[] result = new byte[headerBytes.length + bodyBytes.length];
            System.arraycopy(headerBytes, 0, result, 0, headerBytes.length);
            System.arraycopy(bodyBytes, 0, result, headerBytes.length, bodyBytes.length);

            return result;
        } catch (Exception e) {
            throw new Errors.SerializationException("Failed to serialize message: " + e.getMessage(), null);
        }
    }

    /**
     * Deserialize binary data to message envelope
     */
    public static Types.MessageEnvelope deserialize(byte[] data) throws Errors.SerializationException {
        if (data.length < WIRE_HEADER_SIZE) {
            throw new Errors.SerializationException("Data too short for wire format", null);
        }

        ByteBuffer buffer = ByteBuffer.wrap(data);
        Types.MessageHeader header = deserializeHeader(buffer);
        byte[] bodyData = new byte[buffer.remaining()];
        buffer.get(bodyData);

        Types.MessageBody body = bodyData.length > 0 ? deserializeBody(bodyData) : null;

        Types.MessageEnvelope envelope = new Types.MessageEnvelope();
        envelope.setHeader(header);
        envelope.setBody(body);
        return envelope;
    }

    /**
     * Check if data is valid wire format
     */
    public static boolean isValid(byte[] data) {
        if (data.length < WIRE_HEADER_SIZE) return false;
        ByteBuffer buffer = ByteBuffer.wrap(data);
        int magic = buffer.getInt();
        return magic == WIRE_MAGIC;
    }

    // ============================================================
    // HEADER SERIALIZATION
    // ============================================================

    private static byte[] serializeHeader(Types.MessageHeader header) {
        ByteBuffer buffer = ByteBuffer.allocate(WIRE_HEADER_SIZE);

        buffer.putInt(WIRE_MAGIC);
        buffer.putInt(WIRE_VERSION);

        Integer typeCode = TYPE_MAP.get(header.getType());
        buffer.putInt(typeCode != null ? typeCode : 0);

        int flags = 0;
        if (header.getSignature() != null) flags |= 0x01;
        if (header.getNonce() != null) flags |= 0x02;
        if (header.getReceiver() != null) flags |= 0x04;
        if (header.getCorrelationId() != null) flags |= 0x08;
        if (header.getReplyTo() != null) flags |= 0x10;
        buffer.putInt(flags);

        return buffer.array();
    }

    private static Types.MessageHeader deserializeHeader(ByteBuffer buffer) {
        int magic = buffer.getInt();
        if (magic != WIRE_MAGIC) {
            throw new RuntimeException("Invalid magic number: " + String.format("%08x", magic));
        }

        int version = buffer.getInt();
        int typeCode = buffer.getInt();
        int flags = buffer.getInt();

        Types.MessageType type = TYPE_REVERSE_MAP.getOrDefault(typeCode, Types.MessageType.ERROR);

        Types.MessageHeader header = new Types.MessageHeader();
        header.setType(type);
        header.setVersion(String.format("%d.%d.%d",
                (version >> 24) & 0xFF,
                (version >> 16) & 0xFF,
                (version >> 8) & 0xFF));
        header.setMessageId(UUID.randomUUID().toString());
        header.setTimestamp(Instant.now());
        header.setSender("");
        header.setTtlSeconds(300);
        header.setPriority(0);
        header.setFlags(new ArrayList<>());

        if ((flags & 0x01) != 0) header.setSignature("placeholder");
        if ((flags & 0x02) != 0) header.setNonce("placeholder");
        if ((flags & 0x04) != 0) header.setReceiver("placeholder");
        if ((flags & 0x08) != 0) header.setCorrelationId("placeholder");
        if ((flags & 0x10) != 0) header.setReplyTo("placeholder");

        return header;
    }

    // ============================================================
    // BODY SERIALIZATION
    // ============================================================

    private static byte[] serializeBody(Types.MessageBody body) throws JsonProcessingException {
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("content_type", body.getContentType());
        data.put("data", serializeValue(body.getData()));
        data.put("metadata", body.getMetadata() != null ? body.getMetadata() : new HashMap<>());
        if (body.getSchemaUri() != null) {
            data.put("schema_uri", body.getSchemaUri());
        }

        String json = objectMapper.writeValueAsString(data);
        return json.getBytes(StandardCharsets.UTF_8);
    }

    private static Types.MessageBody deserializeBody(byte[] data) throws Errors.SerializationException {
        try {
            Map<String, Object> parsed = objectMapper.readValue(data, Map.class);

            Types.MessageBody body = new Types.MessageBody();
            body.setContentType((String) parsed.getOrDefault("content_type", "application/json"));
            body.setMetadata((Map<String, Object>) parsed.getOrDefault("metadata", new HashMap<>()));
            body.setSchemaUri((String) parsed.get("schema_uri"));
            body.setSize(data.length);

            Object dataValue = parsed.get("data");
            body.setData(deserializeValue(dataValue));

            return body;
        } catch (Exception e) {
            throw new Errors.SerializationException("Failed to deserialize body: " + e.getMessage(), null);
        }
    }

    // ============================================================
    // VALUE SERIALIZATION
    // ============================================================

    @SuppressWarnings("unchecked")
    private static Object serializeValue(Types.VireoValue<?> value) {
        if (value == null) return null;

        switch (value.getType()) {
            case NULL:
                return null;
            case BOOLEAN:
            case INTEGER:
            case FLOAT:
            case STRING:
                return value.getValue();
            case BINARY:
                byte[] bytes = (byte[]) value.getValue();
                return Map.of("__type", "binary", "value", bytesToHex(bytes));
            case ARRAY:
                List<Types.VireoValue<?>> arr = (List<Types.VireoValue<?>>) value.getValue();
                List<Object> result = new ArrayList<>();
                for (Types.VireoValue<?> v : arr) {
                    result.add(serializeValue(v));
                }
                return result;
            case OBJECT:
                Map<String, Types.VireoValue<?>> obj = (Map<String, Types.VireoValue<?>>) value.getValue();
                Map<String, Object> objResult = new LinkedHashMap<>();
                for (Map.Entry<String, Types.VireoValue<?>> entry : obj.entrySet()) {
                    objResult.put(entry.getKey(), serializeValue(entry.getValue()));
                }
                return objResult;
            case TIMESTAMP:
                Instant ts = (Instant) value.getValue();
                return Map.of("__type", "timestamp", "value", ts.toString());
            case DID:
                return Map.of("__type", "did", "value", value.getValue());
            case SIGNATURE:
                byte[] sig = (byte[]) value.getValue();
                return Map.of("__type", "signature", "value", bytesToHex(sig));
            case NONCE:
                byte[] nonce = (byte[]) value.getValue();
                return Map.of("__type", "nonce", "value", bytesToHex(nonce));
            default:
                return value.getValue();
        }
    }

    @SuppressWarnings("unchecked")
    private static Types.VireoValue<?> deserializeValue(Object data) {
        if (data == null) {
            return new Types.VireoValue<>(Types.VireoType.NULL, null);
        }

        if (data instanceof Boolean) {
            return new Types.VireoValue<>(Types.VireoType.BOOLEAN, data);
        }

        if (data instanceof Integer) {
            return new Types.VireoValue<>(Types.VireoType.INTEGER, data);
        }

        if (data instanceof Double) {
            double d = (Double) data;
            if (d == (int) d) {
                return new Types.VireoValue<>(Types.VireoType.INTEGER, (int) d);
            }
            return new Types.VireoValue<>(Types.VireoType.FLOAT, data);
        }

        if (data instanceof String) {
            return new Types.VireoValue<>(Types.VireoType.STRING, data);
        }

        if (data instanceof List) {
            List<?> list = (List<?>) data;
            List<Types.VireoValue<?>> result = new ArrayList<>();
            for (Object item : list) {
                result.add(deserializeValue(item));
            }
            return new Types.VireoValue<>(Types.VireoType.ARRAY, result);
        }

        if (data instanceof Map) {
            Map<String, Object> map = (Map<String, Object>) data;

            // Check for special types
            if (map.containsKey("__type")) {
                String type = (String) map.get("__type");
                String value = (String) map.get("value");
                switch (type) {
                    case "binary":
                        return new Types.VireoValue<>(Types.VireoType.BINARY, hexToBytes(value));
                    case "timestamp":
                        return new Types.VireoValue<>(Types.VireoType.TIMESTAMP, Instant.parse(value));
                    case "did":
                        return new Types.VireoValue<>(Types.VireoType.DID, value);
                    case "signature":
                        return new Types.VireoValue<>(Types.VireoType.SIGNATURE, hexToBytes(value));
                    case "nonce":
                        return new Types.VireoValue<>(Types.VireoType.NONCE, hexToBytes(value));
                }
            }

            // Regular object
            Map<String, Types.VireoValue<?>> obj = new LinkedHashMap<>();
            for (Map.Entry<String, Object> entry : map.entrySet()) {
                obj.put(entry.getKey(), deserializeValue(entry.getValue()));
            }
            return new Types.VireoValue<>(Types.VireoType.OBJECT, obj);
        }

        return new Types.VireoValue<>(Types.VireoType.NULL, null);
    }

    // ============================================================
    // UTILITY FUNCTIONS
    // ============================================================

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    private static byte[] hexToBytes(String hex) {
        if (hex == null || hex.isEmpty()) return new byte[0];
        int len = hex.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(hex.charAt(i), 16) << 4)
                    + Character.digit(hex.charAt(i + 1), 16));
        }
        return data;
    }
}

File: sdk/python/typescript/src/wire_format.ts
/**
 * Vireo TypeScript SDK - Wire Format Serialization
 * 
 * Binary canonical format for Vireo messages
 * 
 * @module wire_format
 */

import { VireoValue, VireoType, MessageEnvelope, MessageHeader, MessageBody } from './types';
import { SerializationError } from './errors';

// ============================================================
// Wire Format Constants
// ============================================================

export const WIRE_MAGIC = 0x56495245; // 'VIRE' in hex
export const WIRE_VERSION = 0x03000000; // v3.0.0
export const WIRE_HEADER_SIZE = 16; // bytes

// ============================================================
// Wire Format Types
// ============================================================

export interface WireHeader {
  magic: number;
  version: number;
  messageType: number;
  flags: number;
  bodyLength: number;
}

// ============================================================
// Wire Format Serializer
// ============================================================

export class WireFormat {
  /**
   * Serialize a message envelope to binary format
   */
  static serialize(envelope: MessageEnvelope): Uint8Array {
    const headerBytes = this.serializeHeader(envelope.header);
    const bodyBytes = envelope.body ? this.serializeBody(envelope.body) : new Uint8Array(0);

    const result = new Uint8Array(headerBytes.length + bodyBytes.length);
    result.set(headerBytes, 0);
    result.set(bodyBytes, headerBytes.length);

    return result;
  }

  /**
   * Deserialize binary data to message envelope
   */
  static deserialize(data: Uint8Array): MessageEnvelope {
    if (data.length < WIRE_HEADER_SIZE) {
      throw new SerializationError('Data too short for wire format');
    }

    const header = this.deserializeHeader(data);
    const bodyStart = WIRE_HEADER_SIZE;
    const bodyEnd = bodyStart + header.bodyLength;

    if (data.length < bodyEnd) {
      throw new SerializationError('Data too short for body');
    }

    const bodyData = data.slice(bodyStart, bodyEnd);
    const body = this.deserializeBody(bodyData, header.messageType);

    return { header, body };
  }

  /**
   * Serialize message header
   */
  static serializeHeader(header: MessageHeader): Uint8Array {
    const view = new DataView(new ArrayBuffer(WIRE_HEADER_SIZE));

    // Magic number (4 bytes)
    view.setUint32(0, WIRE_MAGIC, false);

    // Version (4 bytes)
    view.setUint32(4, WIRE_VERSION, false);

    // Message type (4 bytes)
    const typeMap: Record<string, number> = {
      'propose': 1,
      'commit': 2,
      'execute': 3,
      'verify': 4,
      'escalate': 5,
      'done': 6,
      'failed': 7,
      'timeout': 8,
      'ack': 9,
      'nack': 10,
      'query': 11,
      'response': 12,
      'error': 13,
    };
    view.setUint32(8, typeMap[header.type] || 0, false);

    // Flags (4 bytes)
    let flags = 0;
    if (header.signature) flags |= 0x01;
    if (header.nonce) flags |= 0x02;
    if (header.receiver) flags |= 0x04;
    if (header.correlationId) flags |= 0x08;
    if (header.replyTo) flags |= 0x10;
    view.setUint32(12, flags, false);

    return new Uint8Array(view.buffer);
  }

  /**
   * Deserialize message header
   */
  static deserializeHeader(data: Uint8Array): MessageHeader {
    const view = new DataView(data.buffer, data.byteOffset);

    const magic = view.getUint32(0, false);
    if (magic !== WIRE_MAGIC) {
      throw new SerializationError(`Invalid magic number: ${magic.toString(16)}`);
    }

    const version = view.getUint32(4, false);
    const typeCode = view.getUint32(8, false);
    const flags = view.getUint32(12, false);

    const typeMap: Record<number, string> = {
      1: 'propose',
      2: 'commit',
      3: 'execute',
      4: 'verify',
      5: 'escalate',
      6: 'done',
      7: 'failed',
      8: 'timeout',
      9: 'ack',
      10: 'nack',
      11: 'query',
      12: 'response',
      13: 'error',
    };

    return {
      type: (typeMap[typeCode] || 'error') as any,
      version: `${(version >> 24) & 0xFF}.${(version >> 16) & 0xFF}.${(version >> 8) & 0xFF}`,
      messageId: crypto.randomUUID(),
      timestamp: new Date(),
      sender: '',
      receiver: (flags & 0x04) ? 'placeholder' : undefined,
      correlationId: (flags & 0x08) ? 'placeholder' : undefined,
      replyTo: (flags & 0x10) ? 'placeholder' : undefined,
      nonce: (flags & 0x02) ? 'placeholder' : undefined,
      signature: (flags & 0x01) ? 'placeholder' : undefined,
      ttlSeconds: 300,
      priority: 0,
      flags: [],
    };
  }

  /**
   * Serialize message body
   */
  static serializeBody(body: MessageBody): Uint8Array {
    const json = JSON.stringify({
      contentType: body.contentType,
      data: this.serializeValue(body.data),
      metadata: body.metadata,
      schemaUri: body.schemaUri,
    });
    return new TextEncoder().encode(json);
  }

  /**
   * Deserialize message body
   */
  static deserializeBody(data: Uint8Array, messageType: number): MessageBody | undefined {
    if (data.length === 0) return undefined;

    const json = new TextDecoder().decode(data);
    const parsed = JSON.parse(json);

    return {
      contentType: parsed.contentType || 'application/json',
      data: this.deserializeValue(parsed.data),
      metadata: parsed.metadata || {},
      schemaUri: parsed.schemaUri,
      size: data.length,
    };
  }

  /**
   * Serialize Vireo value
   */
  static serializeValue(value: VireoValue): any {
    if (value === null || value === undefined) {
      return { __type: 'null', value: null };
    }

    switch (value.type) {
      case 'null':
        return null;
      case 'boolean':
        return value.value;
      case 'integer':
        return value.value;
      case 'float':
        return value.value;
      case 'string':
        return value.value;
      case 'binary':
        return { __type: 'binary', value: Buffer.from(value.value).toString('hex') };
      case 'array':
        return value.value.map((v: any) => this.serializeValue(v));
      case 'object':
        const obj: Record<string, any> = {};
        for (const [k, v] of Object.entries(value.value)) {
          obj[k] = this.serializeValue(v);
        }
        return obj;
      case 'timestamp':
        return { __type: 'timestamp', value: value.value.toISOString() };
      case 'did':
        return { __type: 'did', value: value.value };
      case 'signature':
        return { __type: 'signature', value: Buffer.from(value.value).toString('hex') };
      case 'nonce':
        return { __type: 'nonce', value: Buffer.from(value.value).toString('hex') };
      default:
        return value.value;
    }
  }

  /**
   * Deserialize Vireo value
   */
  static deserializeValue(data: any): VireoValue {
    if (data === null || data === undefined) {
      return { type: 'null', value: null };
    }

    if (typeof data === 'boolean') {
      return { type: 'boolean', value: data };
    }

    if (typeof data === 'number') {
      if (Number.isInteger(data)) {
        return { type: 'integer', value: data };
      }
      return { type: 'float', value: data };
    }

    if (typeof data === 'string') {
      return { type: 'string', value: data };
    }

    if (Array.isArray(data)) {
      return { type: 'array', value: data.map((v) => this.deserializeValue(v)) };
    }

    if (typeof data === 'object' && data !== null) {
      if (data.__type) {
        switch (data.__type) {
          case 'binary':
            return { type: 'binary', value: Buffer.from(data.value, 'hex') };
          case 'timestamp':
            return { type: 'timestamp', value: new Date(data.value) };
          case 'did':
            return { type: 'did', value: data.value };
          case 'signature':
            return { type: 'signature', value: Buffer.from(data.value, 'hex') };
          case 'nonce':
            return { type: 'nonce', value: Buffer.from(data.value, 'hex') };
        }
      }

      const obj: Record<string, VireoValue> = {};
      for (const [k, v] of Object.entries(data)) {
        obj[k] = this.deserializeValue(v);
      }
      return { type: 'object', value: obj };
    }

    return { type: 'null', value: null };
  }

  /**
   * Check if data is valid wire format
   */
  static isValid(data: Uint8Array): boolean {
    try {
      if (data.length < WIRE_HEADER_SIZE) return false;
      const view = new DataView(data.buffer, data.byteOffset);
      const magic = view.getUint32(0, false);
      return magic === WIRE_MAGIC;
    } catch {
      return false;
    }
  }
}